homeassistant/components/media_player/songpal.py

```diff
@@ -0,0 +1,242 @@
+            SongpalDevice(discovery_info["name"],
+                          discovery_info["properties"]["endpoint"]))
+    else:
+        songpal = SongpalDevice(config.get(CONF_NAME),
+                                config.get(CONF_ENDPOINT))
+        devices.append(songpal)
+
+    async_add_devices(devices, True)
+
+    async def async_service_handler(service):
+        """Service handler."""
+        params = {key: value for key, value in service.data.items()
+                  if key != ATTR_ENTITY_ID}
+        entity_id = service.data.get("entity_id", None)
+        _LOGGER.debug("Calling %s (entity: %s) with params %s",
+                      service, entity_id, params)
+
+        for device in devices:
+            if entity_id is None or device.entity_id == entity_id:
+                await device.async_set_sound_setting(params[PARAM_NAME],
+                                                     params[PARAM_VALUE])
+
+    schema = vol.Schema({vol.Required(PARAM_NAME): cv.string,
+                         vol.Required(PARAM_VALUE): cv.string})
+    hass.services.async_register(
+        DOMAIN, SET_SOUND_SETTING, async_service_handler,
+        schema=schema)
+
+
+class SongpalDevice(MediaPlayerDevice):
+    """Class representing a Songpal device."""
+
+    def __init__(self, name, endpoint):
+        """Init."""
+        import songpal
+        self._name = name
+        self.endpoint = endpoint
+        self._dev = songpal.Protocol(self.endpoint)
+        self._sysinfo = None  # type: songpal.Sysinfo
+
+        self._state = False
+        self._available = False
+        self._initialized = False
+
+        self._volume_control = None
+        self._volume_min = 0
+        self._volume_max = 1
+        self._volume = 0
+        self._is_muted = False
+
+        self._sources = []
+
+    @property
+    def name(self):
+        """Return name of the device."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Return an unique ID."""
+        return self._sysinfo.macAddr
+
+    @property
+    def available(self):
+        """Return availability of the device."""
+        return self._available
+
+    @property
+    def dev(self):
+        """Property for accessing the device handle."""
+        return self._dev
+
+    async def async_set_sound_setting(self, name, value):
+        """Change a setting on the device."""
+        res = await self.dev.set_sound_settings(name, value)
+
+        return res
+
+    async def async_update(self):
+        """Fetch updates from the device."""
+        from songpal import SongpalException
+        if not self._initialized:
+            try:
+                await self._dev.get_supported_methods()
+                self._sysinfo = await self.dev.get_system_info()
+                self._initialized = True
+            except SongpalException as ex:
+                _LOGGER.error("Unable to get methods from songpal: %s", ex)
+                raise PlatformNotReady
```

Comment:
This exception only works when called from `async_setup_platform`

Reply:
Ok, that was part of the async_setup_platform earlier, but had to be moved iirc to make unique_id to function. Should the original exception just be raised here instead?

Reply:
Yes, that's the only place where it can get caught. Consider doing the initialization inside setup_platform.

Reply:
IIRC I moved it out for some unique_id related reason. Now the initialization is done again in `async_setup_platform`.